{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "conventional-sculpture",
   "metadata": {},
   "source": [
    "# Catalog Notebook: Overview\n",
    "------\n",
    "This notebook is intended to provide guidance for querying an OpenCGA server through *pyopencga* to explore studies which the user has access to, Clinical data provided in the study (Samples, Individuals Genotypes etc.) and other types of metadata, like permissions.\n",
    "\n",
    "A good first step when start working with OpenCGA is to retrieve information about our user, which projects and studies are we allowed to see.<br>\n",
    "It is also recommended to get a taste of the clinical data we are encountering in the study: How many samples and individuals does the study have? Is there any defined cohorts? Can we get some statistics about the genotypes of the samples in the Sudy?\n",
    "\n",
    "For guidance on how to loggin and get started with *opencga* you can refer to : [pyopencga_first_steps.ipynb](https://github.com/opencb/opencga/tree/develop/opencga-client/src/main/python/notebooks/user-training)\n",
    " \n",
    " **[NOTE]** The server methods used by *pyopencga* client are defined in the following swagger URL:\n",
    "- https://ws.opencb.org/opencga-prod/webservices/   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "random-civilization",
   "metadata": {},
   "source": [
    "## Setup the Client and Login into *pyopencga* \n",
    "\n",
    "**Configuration and Credentials** \n",
    "\n",
    "Let's assume we already have *pyopencga* installed in our python setup (all the steps described on [pyopencga_first_steps.ipynb](https://github.com/opencb/opencga/tree/develop/opencga-client/src/main/python/notebooks/user-training)).\n",
    "\n",
    "You need to provide **at least** a host server URL in the standard configuration format for OpenCGA as a python dictionary or in a json file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "indoor-webster",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logged succesfuly to https://ws.opencb.org/opencga-prod, your token is: eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJkZW1vdXNlciIsImF1ZCI6Ik9wZW5DR0EgdXNlcnMiLCJpYXQiOjE2MTg0Nzg4MDMsImV4cCI6MTYxODQ4MjQwM30.fohz3sZXwIB--s8Bd84vFWxM_RIW6d4mTTGCQjOfazg well done!\n"
     ]
    }
   ],
   "source": [
    "## Step 1. Import pyopencga dependecies\n",
    "from pyopencga.opencga_config import ClientConfiguration # import configuration module\n",
    "from pyopencga.opencga_client import OpencgaClient # import client module\n",
    "from pprint import pprint\n",
    "from IPython.display import JSON\n",
    "# import json\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import datetime\n",
    "\n",
    "## Step 2. OpenCGA host\n",
    "host = 'https://ws.opencb.org/opencga-prod'\n",
    "# host = 'http://localhost:1234/opencga'\n",
    "\n",
    "## Step 3. User credentials\n",
    "user = 'demouser'\n",
    "passwd = 'demouser' ## you can skip this, see below.\n",
    "####################################\n",
    "\n",
    "## Step 4. Create the ClientConfiguration dict\n",
    "config_dict = {'rest': {\n",
    "                       'host': host \n",
    "                    }\n",
    "               }\n",
    "\n",
    "## Step 5. Create the ClientConfiguration and OpenCGA client\n",
    "config = ClientConfiguration(config_dict)\n",
    "oc = OpencgaClient(config)\n",
    "\n",
    "## Step 6. Login to OpenCGA using the OpenCGA client \n",
    "# Pass the credentials to the client\n",
    "# (here we put only the user in order to be asked for the password interactively)\n",
    "# oc.login(user)\n",
    "\n",
    "# or you can pass the user and passwd\n",
    "oc.login(user, passwd)\n",
    "\n",
    "print('Logged succesfuly to {}, your token is: {} well done!'.format(host, oc.token))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "assisted-dakota",
   "metadata": {},
   "source": [
    "## Setup OpenCGA Variables\n",
    "\n",
    "Once we have defined a variable with the client configuration and credentials, we can access to all the methods defined for the client. These methods implement calls to query different data models in *OpenCGA*. \n",
    "\n",
    "Over the user case addressed in this notebook we will be performing queries to the **users, projects, studies, samples, individuals and cohorts**<br> *OpenCGA* data models."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prostate-cancellation",
   "metadata": {},
   "source": [
    "#  Use Cases \n",
    "------\n",
    "\n",
    "In this seciton we are going to show how to work with some of the most common scenarios.<br>\n",
    "- The user-cases addresed here constute a high-level introduction aimed to provide a basis for the user to make their own explorations. \n",
    "- The examples can be adapted to each individual user-case.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "becoming-perception",
   "metadata": {},
   "source": [
    "#  Exploring User Account: Permissios, Projects and Studies\n",
    "------------\n",
    "\n",
    "In this use case we cover retrieving information for our user.\n",
    "\n",
    "**In OpenCGA, all the user permissions are established at a study level**. One project contains **at least** one study, although it may contain several.\n",
    "\n",
    "#### Full Qualified Name (fqn) of Studies\n",
    "It is also very important to understand that in OpenCGA, the projects and studies have a full qualified name (**fqn**) with the format:<br>\n",
    "`[[owner]@[project]]:[study]`\n",
    "\n",
    "We cannot be sure if there might be **other studies** with the same name contained in **other projects**.<br> (E.g: the study *platinium* might be defined in two different projects: *GRch37_project and GRch38_project*)\n",
    "\n",
    "Because of that that, it is recomended to use the **fqn** when referencing studies.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collective-dylan",
   "metadata": {},
   "source": [
    "## 1. Exploring Projects and Studies with our user"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "passing-grenada",
   "metadata": {},
   "source": [
    "### Users: owner and members \n",
    "Depending on the permissions granted, a user can be the owner of a study or just have access to some studies owned by other users.<br>We can retrieve information about our user and its permissions by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "environmental-storage",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using print_results() function:\n",
      "------------------------------------\n",
      "#Time: 14\n",
      "#Num matches: -1\n",
      "#Num results: 1\n",
      "#Num inserted: 0\n",
      "#Num updated: 0\n",
      "#Num deleted: 0\n",
      "#id\taccount.type\tprojects\n",
      "demouser\tGUEST\t.\n",
      "\n",
      "User info using REST response API:\n",
      "id:demouser\taccount_type: GUEST\t projects_owned: 0\n",
      "\n",
      "We can appreciate that our user: demouser has 0 projects from its own: []\n"
     ]
    }
   ],
   "source": [
    "## Getting user information\n",
    "## [NOTE] User needs the quey_id string directly --> (user)\n",
    "#Print using the print_results() function:\n",
    "user_info_resp = oc.users.info(user)\n",
    "user_info_resp.print_results(\"id,account.type,projects\", title='Using print_results() function:') #, metadata=False\n",
    "\n",
    "# Using REST response API:\n",
    "print(\"\\nUser info using REST response API:\")\n",
    "user_info = oc.users.info(user).get_result(0)\n",
    "user_id = user_info['id']  # Define  our user_id\n",
    "user_projects = user_info['projects']  # Define projects owned by our user\n",
    "print('id:{}\\taccount_type: {}\\t projects_owned: {}'.format(user_id, user_info['account']['type'], len(user_projects)))\n",
    "\n",
    "print('\\nWe can appreciate that our user: {} has {} projects from its own: {}'.format(user_id, len(user_projects), user_projects))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "personal-chosen",
   "metadata": {},
   "source": [
    "However, the user might has been granted access to some projects from other users. Let's see how to find this out:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "essential-learning",
   "metadata": {},
   "source": [
    "### Projects:\n",
    "We can list our user's projects using **project client** `search()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "biblical-visibility",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Projects our user (demouser) has access to:\n",
      "------------------------------------------------\n",
      "#id\tname\torganism.scientificName\torganism.assembly\tfqn\n",
      "family\tFamily Studies GRCh37\tHomo sapiens\tGRCh37\tdemo@family\n",
      "population\tPopulation Studies GRCh38\tHomo sapiens\tGRCh38\tdemo@population\n"
     ]
    }
   ],
   "source": [
    "## Getting user projects\n",
    "## [NOTE] Client specific methods have the query_id as a key:value (i.e (user=user_id)) \n",
    "\n",
    "projects_info = oc.projects.search()\n",
    "projects_info.print_results(fields='id,name,organism.scientificName,organism.assembly,fqn', title='Projects our user ({}) has access to:'.format(user_id), metadata=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "restricted-shelter",
   "metadata": {},
   "source": [
    "The **fqn** `owner@project` shows the owner of the project/s; this owner has granted permission to our user to the projects above.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "single-float",
   "metadata": {},
   "source": [
    "### Studies:\n",
    "Let's see which studies do we have access within the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "every-township",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For this user-case, we can use project:family\n"
     ]
    }
   ],
   "source": [
    "# First we define one projectId\n",
    "project_info = oc.projects.search().get_result(0)\n",
    "project_id = project_info['id']\n",
    "\n",
    "print('For this user-case, we can use project:{}'.format(project_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "gentle-gothic",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Our user (user_id) has access to 2 different studies within the family project\n",
      "\n",
      "project:family\t study_id:platinum\t study_fqn:demo@family:platinum \n",
      "project:family\t study_id:corpasome\t study_fqn:demo@family:corpasome \n"
     ]
    }
   ],
   "source": [
    "studies = oc.studies.search(project_id)\n",
    "\n",
    "## Print the studies using the result_iterator() method\n",
    "print('Our user (user_id) has access to 2 different studies within the {} project\\n'.format(project_id))\n",
    "for study in studies.result_iterator():\n",
    "    print(\"project:{}\\t study_id:{}\\t study_fqn:{} \".format(project_id, study['id'], study['fqn']))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "interesting-terrorist",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Let's use the study: platinum with fqn: demo@family:platinum\n"
     ]
    }
   ],
   "source": [
    "# Define the study we are going to work with\n",
    "study_info = oc.studies.search(project_id).get_result(0)\n",
    "study_id = study_info['id']\n",
    "study_fqn = study_info['fqn']\n",
    "print(\"Let's use the study: {} with fqn: {}\".format(study_id, study_fqn))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "funded-policy",
   "metadata": {},
   "source": [
    "## 2. Checking Groups and Permissions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prime-christian",
   "metadata": {},
   "source": [
    "Now we can assume that we want to check to which groups our user belongs to and which permisions pur user has been granted for the study (remember that all the permissions are established at the study level).<br>\n",
    "- The first step might be check which groups exist within the **study**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "changing-friendship",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "group_id: @members\n",
      "group_id: @admins\n",
      "group_id: @opencb-team\n",
      "\n",
      "There are 3 groups in the study demo@family:platinum: ['@members', '@admins', '@opencb-team']\n"
     ]
    }
   ],
   "source": [
    "# Query to the study web service\n",
    "groups = oc.studies.groups(study_fqn)\n",
    "study_groups = []  # Define an empty list for the groups\n",
    "\n",
    "## This will give us the whole list of groups existing in the study\n",
    "for group in groups.result_iterator():\n",
    "    study_groups.append(group['id'])\n",
    "    print(\"group_id: {}\".format(group['id']))\n",
    "    \n",
    "print('\\nThere are 3 groups in the study {}: {}'.format(study_fqn, study_groups))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satisfied-florida",
   "metadata": {},
   "source": [
    "- If we want to check in which groups is our user included:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "mobile-report",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "group_id: @members\n",
      "\n",
      "Our user demouser belongs to group/s: ['@members']\n"
     ]
    }
   ],
   "source": [
    "user_groups = [] # Define an empty list \n",
    "\n",
    "## This will give us only the groups our user belongs to\n",
    "for group in groups.result_iterator():\n",
    "    if user_id in group['userIds']:\n",
    "        user_groups.append(group['id'])\n",
    "        print(\"group_id: {}\".format(group['id']))\n",
    "        \n",
    "print('\\nOur user {} belongs to group/s: {}'.format(user_id, user_groups))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dental-specialist",
   "metadata": {},
   "source": [
    "Independently of the groups defined for a study, our user always belongs to the group **members**, which is one of the default groups in *OpenCGA*."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "optional-girlfriend",
   "metadata": {},
   "source": [
    "- We might be wondering which specific permissions our user has. We can check this using the `client.acl()` method (**acl** = access control list):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "according-spare",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The user demouser  has the following permissions:\n",
      "\n",
      " ['VIEW_PANELS', 'VIEW_FAMILIES', 'VIEW_JOBS', 'VIEW_FILES', 'VIEW_FILE_ANNOTATIONS', 'VIEW_COHORTS', 'VIEW_SAMPLE_VARIANTS', 'VIEW_FAMILY_ANNOTATIONS', 'VIEW_FILE_HEADER', 'VIEW_FILE_CONTENT', 'VIEW_AGGREGATED_VARIANTS', 'VIEW_INDIVIDUALS', 'VIEW_COHORT_ANNOTATIONS', 'VIEW_SAMPLES', 'VIEW_SAMPLE_ANNOTATIONS', 'VIEW_CLINICAL_ANALYSIS', 'EXECUTE_JOBS', 'VIEW_INDIVIDUAL_ANNOTATIONS']\n"
     ]
    }
   ],
   "source": [
    "# Permissions granted directly to user:\n",
    "acls = oc.studies.acl(study_id, member=user_id).get_result(0)\n",
    "print('The user',user_id,' has the following permissions:\\n\\n', acls[user_id])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "piano-adobe",
   "metadata": {},
   "source": [
    "The default groups in *OpenCGA* are: **members** and **admins**. We can also check the permissions granted to **members** directly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "classical-understanding",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "group_id: @members\t group_acls: []\n"
     ]
    }
   ],
   "source": [
    "# Permissions granted to the groups our user belongs to\n",
    "for group in user_groups:\n",
    "    acls = oc.studies.acl(study_id, member=group)\n",
    "    if acls.get_num_results() == 0:\n",
    "        print('group_id: {}\\t group_acls: []'.format(group))\n",
    "    else:\n",
    "        for acl in acls.result_iterator():\n",
    "            print('group_id: {}\\t group_acls: {}'.format(group, acl[group]))\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dramatic-sixth",
   "metadata": {},
   "source": [
    "Intuitively, the group **members** is the basic group and has any default permissions. On the other hand, users in the group **admins** have permission to see and edit the study information.\n",
    "\n",
    "For more information about user and group permissions, check the official *OpenCGA* documentation: **[Catalog and Security - Users and Permissions](http://docs.opencb.org/display/opencga/Sharing+and+Permissions)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "soviet-steal",
   "metadata": {},
   "source": [
    "# Exploring Catalog Clinical Metadata\n",
    "-----------------------\n",
    "A genomic data analysis platform need to keep track of different resources such as:\n",
    "\n",
    "- Clinical Data: information about individuals, samples from those individuals etc.\n",
    "- Files Metadata: information about files contained in the platform, such as VCFs and BAMs.\n",
    "\n",
    "*OpenCGA Catalog* is the component that assumes this role by storing this kind of information"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "soviet-logan",
   "metadata": {},
   "source": [
    "## 1. Exploring Samples and Individuals\n",
    "\n",
    "Once we know the studies our user has access to, we can explore the samples within the study.<br>\n",
    "To fetch samples you need to use the sample client built in **pyopencga**. Remember that it is recomended to use the **[fqn](#Full-Qualified-Name-(fqn)-of-Studies )** when referencing studies.<br>\n",
    "\n",
    "### Samples:\n",
    "Let's imagine we want to know how many samples are in the **study** stored in the `study_fqn` variable, and list information about the first two samples: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "nuclear-climate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info from 5 samples from study demo@family:platinum\n",
      "--------------------------------------------------------\n",
      "#Time: 68\n",
      "#Num matches: 17\n",
      "#Num results: 5\n",
      "#Num inserted: 0\n",
      "#Num updated: 0\n",
      "#Num deleted: 0\n",
      "#id\tcreationDate\tsomatic\tphenotypes.id\tphenotypes.name\tindividualId\n",
      "NA12877\t20200625131818\tFalse\t.\t.\tNA12877\n",
      "NA12878\t20200625131819\tFalse\t.\t.\tNA12878\n",
      "NA12879\t20200625131820\tFalse\t.\t.\tNA12879\n",
      "NA12880\t20200625131821\tFalse\t.\t.\tNA12880\n",
      "NA12881\t20200625131822\tFalse\t.\t.\tNA12881\n"
     ]
    }
   ],
   "source": [
    "## Call to the sample web endpoint\n",
    "samples = oc.samples.search(study=study_fqn, includeIndividual=True, count=True, limit = 5) ## other possible params, count=False, id='NA12880,NA12881'\n",
    "samples.print_results(fields='id,creationDate,somatic,phenotypes.id,phenotypes.name,individualId', title='Info from 5 samples from study {}'.format(study_fqn))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer\n",
    "#JSON(samples.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "descending-shipping",
   "metadata": {},
   "source": [
    "We can see that the number of samples in the study is given by **#Num matches**. The count is returned because we have set the parameter `count=True`.\n",
    "\n",
    "However, we have used the parameter `limit` to restrict the number of samples the query returns. We can get all the samples ids by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "fundamental-suggestion",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 17 samples with ids:\n",
      " ['NA12877', 'NA12878', 'NA12879', 'NA12880', 'NA12881', 'NA12882', 'NA12883', 'NA12884', 'NA12885', 'NA12886', 'NA12887', 'NA12888', 'NA12889', 'NA12890', 'NA12891', 'NA12892', 'NA12893']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sample_ids = [] # Define an empty list\n",
    "\n",
    "# Define a new sample query without limit\n",
    "samples = oc.samples.search(study=study_fqn, count=True) \n",
    "\n",
    "for sample in samples.result_iterator():\n",
    "    sample_ids.append(sample['id'])\n",
    "\n",
    "print('There are {} samples with ids:\\n {}\\n'.format(len(sample_ids), sample_ids))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "muslim-compiler",
   "metadata": {},
   "source": [
    "### Individuals:\n",
    "Now, we can repite the same process for check the number of individuals in the **study** . The difference is that now we will be making a call to the **individuals** web service:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "interstate-helena",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information about 5 individuals in the studydemo@family:platinum\n",
      "---------------------------------------------------------------------\n",
      "#Time: 72\n",
      "#Num matches: 17\n",
      "#Num results: 5\n",
      "#Num inserted: 0\n",
      "#Num updated: 0\n",
      "#Num deleted: 0\n",
      "#id\tsex\n",
      "NA12877\tUNKNOWN\n",
      "NA12878\tUNKNOWN\n",
      "NA12879\tUNKNOWN\n",
      "NA12880\tUNKNOWN\n",
      "NA12881\tUNKNOWN\n"
     ]
    }
   ],
   "source": [
    "## Using the individuals search web service\n",
    "individuals = oc.individuals.search(study=study_fqn, count=True, limit=5) ## other possible params, count=False, id='NA12880,NA12881'\n",
    "individuals.print_results(fields='id,sex', title='Information about 5 individuals in the study{}'.format(study_fqn))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer\n",
    "#JSON(individuals.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "invalid-chase",
   "metadata": {},
   "source": [
    "- We might be interested in knowing when the individuals were added to *OpenCGA*, or the individuals sex. Since **pyopencga 2.0.1.1** it is possible to export the results to a *pandas dataframe* object with the function `to_data_frame()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "rapid-librarian",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        id      sex    creationDate\n",
      "0  NA12877  UNKNOWN  20200625131812\n",
      "1  NA12878  UNKNOWN  20200625131813\n",
      "2  NA12879  UNKNOWN  20200625131813\n",
      "3  NA12880  UNKNOWN  20200625131813\n",
      "4  NA12881  UNKNOWN  20200625131814\n"
     ]
    }
   ],
   "source": [
    "## Using the individuals search web service without limit param\n",
    "individuals = oc.individuals.search(study=study_fqn) \n",
    "## Using the new function to_data_frame()\n",
    "individuals_df = individuals.to_data_frame()\n",
    "print(individuals_df[['id','sex','creationDate']].head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "liquid-decimal",
   "metadata": {},
   "source": [
    "### Custom Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "critical-george",
   "metadata": {},
   "source": [
    "## 2. Exploring Files\n",
    "We can start by exploring the number of files in the study, and retrieveing information about one file as an example of which kind of data is stored in the **file** data model of *OpenCGA*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "retired-mounting",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information about files in study demo@family:platinum\n",
      "----------------------------------------------------------\n",
      "#Time: 94\n",
      "#Num matches: 3874\n",
      "#Num results: 5\n",
      "#Num inserted: 0\n",
      "#Num updated: 0\n",
      "#Num deleted: 0\n",
      "#id\tformat\tsize\tsoftware\n",
      "data:platinum-genomes-vcf-NA12877_S1.genome.vcf.gz\tVCF\t887890738\t{}\n",
      "data:platinum-genomes-vcf-NA12878_S1.genome.vcf.gz\tVCF\t883195909\t{}\n",
      "data:platinum-genomes-vcf-NA12879_S1.genome.vcf.gz\tVCF\t889974818\t{}\n",
      "data:platinum-genomes-vcf-NA12880_S1.genome.vcf.gz\tVCF\t899309868\t{}\n",
      "data:platinum-genomes-vcf-NA12881_S1.genome.vcf.gz\tVCF\t918334187\t{}\n"
     ]
    }
   ],
   "source": [
    "## Using the files web service\n",
    "files = oc.files.search(study=study_fqn, count=True, type='FILE', limit=5, exclude='attributes') ## other possible params, count=False, id='NA12880,NA12881'\n",
    "files.print_results(fields='id,format,size,software', title='Information about files in study {}'.format(study_fqn))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer\n",
    "#JSON(files.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dressed-emission",
   "metadata": {},
   "source": [
    "There is plenty of useful information contained in the file data model like the file format, the stats, size of the file. If we want to look for more concrete information about one specific file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "historical-tourism",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The study demo@family:platinum contains a VCF file with id: data:platinum-genomes-vcf-NA12878_S1.genome.vcf.gz,\n",
      "created on: 20200625131819\n"
     ]
    }
   ],
   "source": [
    "my_vcf = files.get_result(1)\n",
    "print('The study {} contains a {} file with id: {},\\ncreated on: {}'.format(study_fqn, my_vcf['format'], \n",
    "                                                                            my_vcf['id'], my_vcf['creationDate']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "numerous-nirvana",
   "metadata": {},
   "source": [
    "- We can also be interested in knowing the number of files for a specific sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "infinite-venue",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sample NA12877 has file/s: ['data:platinum-genomes-vcf-NA12877_S1.genome.vcf.gz']\n"
     ]
    }
   ],
   "source": [
    "## Using the samples info web service\n",
    "sample_of_interest = sample_ids[0]\n",
    "\n",
    "## List the files for a concrete sample\n",
    "sample = oc.samples.info(study=study_fqn, samples= sample_of_interest) ## other possible params, count=False, id='NA12880,NA12881'\n",
    "sample_files = sample.get_result(0)['fileIds']\n",
    "\n",
    "print('The sample {} has file/s: {}'.format(sample_of_interest, sample_files))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "juvenile-newman",
   "metadata": {},
   "source": [
    "## 3. Exploring Cohorts\n",
    "\n",
    "One powerful feature of *OpenCGA* is the possibility of define **cohorts** that include individuals with common traits of interest, like a phenotype, nationality etc.\n",
    "The **cohorts** are defined at the study level. *OpenCGA* creates a default cohort *ALL*, which includes all the individuals of the study.\n",
    "\n",
    " We can explore which cohorts are defined in the  **study** by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "small-biodiversity",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information about cohorts in study demo@family:platinum\n",
      "------------------------------------------------------------\n",
      "#Time: 42\n",
      "#Num matches: 1\n",
      "#Num results: 1\n",
      "#Num inserted: 0\n",
      "#Num updated: 0\n",
      "#Num deleted: 0\n",
      "#id\ttype\tdescription\tnumSamples\n",
      "ALL\tCOLLECTION\tDefault cohort with almost all indexed samples\t17\n"
     ]
    }
   ],
   "source": [
    "## Using the cohorts search web service\n",
    "cohorts = oc.cohorts.search(study=study_fqn, count=True, exclude='samples') ## other possible params, count=False, id='NA12880,NA12881'\n",
    "cohorts.print_results(fields='id,type,description,numSamples', title='Information about cohorts in study {}'.format(study_fqn))\n",
    "\n",
    "## Uncomment next line to display an interactive JSON viewer\n",
    "#JSON(cohorts.get_results())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impressive-crisis",
   "metadata": {},
   "source": [
    "**[NOTE]**: For any **study** in *OpenCGA* the default cohort **ALL** is always present. As we can see in the description of the cohort data model, \"**ALL** is the default cohort with almost all indexed samples\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "artificial-given",
   "metadata": {},
   "source": [
    "# Aggregations\n",
    "-------\n",
    "You can easily filter samples, individuals, ... using your custom annotation ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "automatic-knight",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}